master/buildbot/steps/artifact.py

```diff
@@ -39,7 +45,76 @@ def forceRebuild(build):
 
     return force_chain_rebuild or force_rebuild
 
-class FindPreviousSuccessfulBuild(ResumeBuild):
+class PreviousBuildStatus:
+    Forced = 1
+    Unmergeable = 2
+    NotFound = 3
+    Found = 4
+
+class FindPreviousSuccessBuildMixin():
+    disableUnmergeable = False
+
+    @defer.inlineCallbacks
+    def _determinePreviousBuild(self, master, build, build_sourcestamps):
+        if forceRebuild(build):
+            defer.returnValue((PreviousBuildStatus.Forced, None))
+
+        mergeRequestFn = build.builder.getConfiguredMergeRequestsFn()
+
+        if mergeRequestFn == False and not self.disableUnmergeable:
+            log.msg("[brid: %d] is forced - skipping previous build" % build.requests[0].id)
+            defer.returnValue((PreviousBuildStatus.Unmergeable, None))
+
+        elif mergeRequestFn == True or mergeRequestFn == False: # Default case, with no merge function assigned
+            log.msg("[brid: %d] is searcing for previous successful build without merge function" % build.requests[0].id)
+            prevBuildRequest = yield master.db.buildrequests \
+                .getBuildRequestBySourcestamps(buildername=build.builder.config.name,
+                                               sourcestamps=build_sourcestamps)
+            if prevBuildRequest:
+                log.msg("[brid: %d] previous successful build [%s] found" % (build.requests[0].id, prevBuildRequest['brid']))
+                defer.returnValue((PreviousBuildStatus.Found, prevBuildRequest))
+
+        else: # Custom merge function assigned
+            log.msg("[brid: %d] is searcing for previous successful build with merge function" % build.requests[0].id)
+            prevBuildRequests = yield master.db.buildrequests \
```

Comment:
@tdctaz we are not breaking execution(returning) after defer.returnValue, it's fine to leave the code as it is